wasi: add support for version when creating WASI

mhdawson opened the pull request:
Refs: https://github.com/nodejs/node/issues/46254

- add version to options when creating WASI object
- add convenience function to return importObject

Signed-off-by: Michael Dawson <mdawson@devrus.com>

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

devsnek commented:
why do we need to explicitly pass a version? each version already has its own import namespace, so they can just coexist right?

mhdawson commented:
@devsnek you can read some of the discussion suggestion for a version in https://github.com/nodejs/node/issues/46254

guybedford reviewed (approved):
Thanks for working on this @mhdawson. Looks good to me and the importObject seems useful, although I was hoping we could discuss some more API changes before removing the flag as well.

guybedford commented on doc/api/wasi.md:
```diff
@@ -16,21 +16,18 @@ import { WASI } from 'wasi';
 import { argv, env } from 'node:process';
 
 const wasi = new WASI({
+  version: 'wasi_snapshot_preview1',
```
At the very least, `snapshot-preview1` may be a bit briefer.

mhdawson replied:
@cjihrig I'd defer to you on how to best name the versions. I just used exactly what was documented as that makes it obvious what it maps to.

cjihrig replied:
@guybedford probably has better access to the folks that are determining the names of the versions. I'd be in favor of at least dropping the "wasi_" prefix though.

mhdawson replied:
Ok, will update to 'preview1'

mhdawson replied:
As I started doing that I realised that the other thing is it means that the mapping between the version requested and what goes into the import object is not as clear.  In terms of the discussion about documenting  `{ wasi_unstable: wasi.wasiImport }` adding it as a version that could be requested would make less sense if we don't have a 1 to 1 mapping between the version requested and the name of the field in the import object.

I'll wait until next week to give a bit more time for you two to comment on that aspect. If we still think usigin `preview1` is better I'll document the `{ wasi_unstable: wasi.wasiImport }` some other way.

mhdawson replied:
Talked with @guybedford seems like for preview2 the binding between the version and the imports is not direct anyway so my reasoning for keeping them in line would no longer make sense. Updating to 'preview1'

cjihrig commented on doc/api/wasi.md:
```diff
@@ -148,6 +146,19 @@ added:
     WebAssembly application. **Default:** `1`.
   * `stderr` {integer} The file descriptor used as standard error in the
     WebAssembly application. **Default:** `2`.
+  * `version` {string} The version of wasi requested. Currently the only
```
```suggestion
  * `version` {string} The version of WASI requested. Currently the only
```

cjihrig commented on doc/api/wasi.md:
```diff
@@ -148,6 +146,19 @@ added:
   * `stderr` {integer} The file descriptor used as standard error in the
     WebAssembly application. **Default:** `2`.
+  * `version` {string} The version of wasi requested. Currently the only
+    supported version is `wasi_snapshot_preview1`
```
```suggestion
    supported version is `wasi_snapshot_preview1`.
```

cjihrig commented on doc/api/wasi.md:
```diff
@@ -148,6 +146,19 @@ added:
+added: REPLACEME
+-->
+
+Return an import object that can be passed to `wasi.start` if
```
The result of this function isn't supposed to be passed to `wasi.start()` is it?

mhdawson replied:
Thanks for catching that.

mhdawson replied:
fixed

cjihrig commented on doc/api/wasi.md:
```diff
@@ -148,6 +146,19 @@ added:
+-->
+
+Return an import object that can be passed to `wasi.start` if
+no other WASM imports are needed beyond those provided by wasi. It
```
```suggestion
no other WASM imports are needed beyond those provided by WASI. It
```

cjihrig commented on doc/api/wasi.md:
```diff
@@ -148,6 +146,19 @@ added:
+
+Return an import object that can be passed to `wasi.start` if
+no other WASM imports are needed beyond those provided by wasi. It
+will reflect the version of wasi requested when `new WASI` was
```
```suggestion
will reflect the version of WASI requested when `new WASI()` was
```

mhdawson replied:
done

cjihrig commented on doc/api/wasi.md:
```diff
@@ -16,21 +16,18 @@ import { WASI } from 'wasi';
   },
 });
 
-// Some WASI binaries require:
-//   const importObject = { wasi_unstable: wasi.wasiImport };
-const importObject = { wasi_snapshot_preview1: wasi.wasiImport };
-
```
We appear to be losing this information, which could be important to some users. I think we either need to keep this somewhere in the docs, or support passing in "wasi_unstable" as a version which would return the same API as wasi_snapshot_preview1, but with a different version string.

mhdawson replied:
I'm happy to either add it back to the doc or add to "wasi_unstable" to what  can be passed in as a version depending on what people think is best.

cjihrig replied:
It seems like a pretty easy thing to support since it would be an alias for preview1, so I'd suggest supporting it and removing this note from the docs (but we would need to document that "wasi_unstable" is a valid version).

mhdawson replied:
will do

cjihrig commented:
> although I was hoping we could discuss some more API changes before removing the flag as well.

@guybedford what else did you have in mind? Maybe open another issue. I'd hate for us to remove the flag and have something slip through the cracks.

mhdawson commented on lib/wasi.js:
```diff
@@ -45,6 +47,25 @@ class WASI {
+      }
+      this[kVersion] = options.version;
+    } else {
+      ({ WASI: _WASI } = internalBinding('wasi'));
```
ack

mhdawson replied:
added

mhdawson commented:
@cjihrig, @guybedford  I think I've addressed all the comments so far.

The one outstanding thing is that the tests are not run for 'unstable' since they were compiled against 'wasi_snapshot_preview1'. I think we would need a new set of the compiled wasm test files. Does that need to be handled in this PR or is it ok since there seems to have been no prior testing of unstable anyway?

guybedford commented:
The `unstable` import interface is very old at this point so I think it is fine to treat that as a legacy path that we can support but that we don't need to explicitly test.

mhdawson commented:
Pushed commit to remove disabled testing related to unstable version. Should be ready to go at this point.

cjihrig reviewed (approved):
LGTM. Thanks for doing this.

mhdawson commented:
Looks like status from CI did not post correct. CI is complete and passed, going to land.

mhdawson commented:
Landed in 2d35f669ae59

juanarbol commented:
Would you mind backporting this to v18.x?
